### PR TITLE
grpc: factor out base classes for gRPC based drivers

### PR DESCRIPTION
**Pull request by alltilla**
TODO:
 - [x] fix macOS build
 - [x] manually try out all the drivers
 - [x] write news entries about the new options
 - [x] follow-up using macros in the parser.h files in cfg-helper

For the reviewers:
 * It is not trivial which options are gRPC related and which are actual driver specific. I have made a classification myself and factored out what I thought to be common, but please think about it yourselves, while you are reviewing.
 * Some features that I classified as gRPC related were not implemented in some drivers, so I added them. Please double check if the newly added features make sense in the drivers.
 * Is it the best to statically link the `grpc-common` target? I thought that dynamic linking would be better but I received a lot of linker warnings about it not being portable. Have I missed something?
 * ~~Is there a better way to merge two sets of keywords than defining a macro like `GRPC_KEYWORDS`? It messes up the cfg-helper :/~~ `gcc -E` and a bit of parsing improvement in `axosyslog-cfg-helper` will be able to handle this.

### CODE REVIEW

File: modules/grpc/common/grpc-dest.cpp [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+  log_threaded_dest_driver_init_instance(&self->super, cfg);
+
+  self->super.super.super.super.init = _init;
+  self->super.super.super.super.deinit = _deinit;
+  self->super.super.super.super.free_fn = _free;
+  self->super.super.super.super.generate_persist_name = _generate_persist_name;
+
+  self->super.worker.construct = _construct_worker;
+  self->super.stats_source = stats_register_type(stats_name); // todo: comes from a virtual fn
```
Comment (alltilla) [inline modules/grpc/common/grpc-dest.cpp:200]: Remove the comment, this is handled by the function argument.

File: modules/grpc/otel/otel-dest-worker.cpp [hunk 2]
```diff
@@ -41,55 +39,22 @@ using namespace opentelemetry::proto::trace::v1;
...
     spans_current_batch_bytes(0),
     formatter(s->super.owner->super.super.super.cfg)
 {
-  ::grpc::ChannelArguments args;
-
-  if (owner.get_compression())
-    {
-      args.SetCompressionAlgorithm(GRPC_COMPRESS_GZIP);
-    }
+  std::shared_ptr<::grpc::ChannelCredentials> credentials = DestWorker::create_credentials(); // todo: error handling?
```
Comment (alltilla) [inline modules/grpc/otel/otel-dest-worker.cpp:49]: Fix this.

File: modules/grpc/loki/loki-worker.h [hunk 3]
```diff
@@ -1,37 +0,0 @@
...
-#define LOKI_WORKER_H
-
-#include "syslog-ng.h"
-
-#include "compat/cpp-start.h"
-#include "logthrdest/logthrdestdrv.h"
-
-typedef struct _LokiDestWorker LokiDestWorker;
-
-LogThreadedDestWorker *loki_dw_new(LogThreadedDestDriver *o, gint worker_index);
```
Comment (alltilla) [inline modules/grpc/loki/loki-worker.h:33]: Belongs to next commit.

File: modules/grpc/common/grpc-dest.cpp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+{
+  credentials_builder_wrapper.self = &credentials_builder;
+}
+
+bool
+DestDriver::init()
+{
+  if (url.length() == 0)
+    {
+      msg_error("OpenTelemetry: url() option is mandatory", // todo: make the descendant class define the beginning of the message
```
Comment (alltilla) [inline modules/grpc/common/grpc-dest.cpp:43]: Fix this.

**Review by OverOrion — COMMENTED** [review]
Let me get back to you with answers to your questions / requests.
Overall a solid looking PR, huge kudos for doing the extraction :100:

File: lib/merge-grammar.py [hunk 5]
```diff
@@ -55,22 +56,31 @@ def print_to_stdout(line):
...
-            if start_marker in line:
-                in_block = True
-            elif end_marker in line:
-                in_block = False
-            elif in_block:
-                print_to_stdout(line)
+            if line.startswith('/* START_') and line.endswith(' */\n'):
+                block_name = line[len('/* START_'):-len(' */\n')]
+            elif block_name and line == '/* END_' + block_name + ' */\n':
+                BLOCKS[block_name] = BLOCKS.get(block_name, '') + ''.join(block)
```
[the remark this answers is not in the document]
Comment (alltilla) [reply lib/merge-grammar.py:67]: It is, and it should append to the previous. It is already built upon in `cfg-grammar.y`.

File: modules/grpc/bigquery/bigquery-worker.cpp [hunk 6]
```diff
@@ -164,7 +151,7 @@ DestinationWorker::should_initiate_flush()
 DestinationWorker::Slice
 DestinationWorker::format_template(LogTemplate *tmpl, LogMessage *msg, GString *value, LogMessageValueType *type)
 {
-  DestinationDriver *owner = this->get_owner();
+  DestinationDriver *owner_ = this->get_owner();
```
[the remark this answers is not in the document]
Comment (alltilla) [reply modules/grpc/bigquery/bigquery-worker.cpp:154]: `owner` would shadow `this->owner`.

File: modules/grpc/credentials/Makefile.am [hunk 7]
```diff
@@ -1,6 +0,0 @@
-grpc_credentials_sources = \
```
[the remark this answers is not in the document]
Comment (alltilla) [reply modules/grpc/credentials/Makefile.am:1]: I did `git mv`.

```
$ git diff --stat 73382fbd9c5778ae5e28649bd837e9d507efe8e9~ 73382fbd9c5778ae5e28649bd837e9d507efe8e9
 modules/grpc/CMakeLists.txt                                        |  3 +--
 modules/grpc/Makefile.am                                           |  3 +--
 modules/grpc/bigquery/CMakeLists.txt                               |  4 ++--
 modules/grpc/bigquery/Makefile.am                                  |  7 ++++---
 modules/grpc/common/CMakeLists.txt                                 | 13 +++++++++++++
 modules/grpc/common/Makefile.am                                    | 33 +++++++++++++++++++++++++++++++++
 modules/grpc/common/credentials/CMakeLists.txt                     |  5 +++++
 modules/grpc/common/credentials/Makefile.am                        |  6 ++++++
 modules/grpc/{ => common}/credentials/grpc-credentials-builder.cpp |  0
 modules/grpc/{ => common}/credentials/grpc-credentials-builder.h   |  0
 modules/grpc/{ => common}/credentials/grpc-credentials-builder.hpp |  0
 modules/grpc/common/metrics/CMakeLists.txt                         |  4 ++++
 modules/grpc/common/metrics/Makefile.am                            |  5 +++++
 modules/grpc/{ => common}/metrics/grpc-metrics.cpp                 |  0
 modules/grpc/{ => common}/metrics/grpc-metrics.hpp                 |  0
 modules/grpc/credentials/CMakeLists.txt                            |  5 -----
 modules/grpc/credentials/Makefile.am                               |  6 ------
 modules/grpc/loki/CMakeLists.txt                                   |  4 ++--
 modules/grpc/loki/Makefile.am                                      |  8 ++++----
 modules/grpc/metrics/CMakeLists.txt                                |  4 ----
 modules/grpc/metrics/Makefile.am                                   |  5 -----
 modules/grpc/otel/CMakeLists.txt                                   |  6 ++----
 modules/grpc/otel/Makefile.am                                      | 13 ++++++-------
 tests/copyright/policy                                             |  3 +--
 24 files changed, 89 insertions(+), 48 deletions(-)
```

Not sure why it is not working in your env.

File: modules/grpc/common/grpc-dest-worker.cpp [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+#include "grpc-dest-worker.hpp"
+
+using namespace syslogng::grpc;
+
+/* C++ Implementations */
+
+DestWorker::DestWorker(GrpcDestWorker *s)
+  : super(s),
+    owner(*((GrpcDestDriver *) s->super.owner)->cpp)
```
[the remark this answers is not in the document]
Comment (alltilla) [reply modules/grpc/common/grpc-dest-worker.cpp:37]: Thanks for the idea. `reinterpret_cast` was what worked.

File: modules/grpc/common/grpc-source.cpp [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+SourceDriver::init()
+{
+  if (!this->port)
+    {
+      msg_error("Failed to initialize gRPC based source, port() must be set",
+                log_pipe_location_tag(&this->super->super.super.super.super));
+      return FALSE;
+    }
+
+  if (!this->credentials_builder.validate())
```
[the remark this answers is not in the document]
Comment (alltilla) [reply modules/grpc/common/grpc-source.cpp:48]: `validate()` logs the specific error inside.

File: modules/grpc/common/grpc-source.cpp [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    return FALSE;
+
+  ::grpc::EnableDefaultHealthCheckService(true);
+
+  this->super->super.worker_options.super.init_window_size /= super->super.num_workers;
+
+  if (fetch_limit == -1)
+    fetch_limit = this->super->super.worker_options.super.init_window_size;
```
[the remark this answers is not in the document]
Comment (alltilla) [reply modules/grpc/common/grpc-source.cpp:56]: `LogThredeadSourceDriver` does not descend from `LogSource`.

File: modules/grpc/common/grpc-dest-worker.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (alltilla) [reply modules/grpc/common/grpc-dest-worker.cpp]: Done.

File: modules/grpc/common/grpc-dest.hpp [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  void add_extra_channel_arg(std::string name, long value)
+  {
+    this->int_extra_channel_args.push_back(std::pair<std::string, long> {name, value});
+  }
+
+  void add_extra_channel_arg(std::string name, std::string value)
+  {
+    this->string_extra_channel_args.push_back(std::pair<std::string, std::string> {name, value});
```
[the remark this answers is not in the document]
Comment (alltilla) [reply modules/grpc/common/grpc-dest.hpp]: Thanks, done.

File: modules/grpc/common/grpc-dest.hpp [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  void add_header(std::string name, std::string value)
+  {
+    std::transform(name.begin(), name.end(), name.begin(), ::tolower);
+    this->headers.push_back(std::pair<std::string, std::string> {name, value});
```
[the remark this answers is not in the document]
Comment (alltilla) [reply modules/grpc/common/grpc-dest.hpp]: Thanks, done.

File: modules/grpc/common/grpc-source.hpp [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  int get_concurrent_requests()
+  {
+    return this->concurrent_requests;
+  }
+
+  void add_extra_channel_arg(std::string name, long value)
+  {
+    this->int_extra_channel_args.push_back(std::pair<std::string, long> {name, value});
```
[the remark this answers is not in the document]
Comment (alltilla) [reply modules/grpc/common/grpc-source.hpp]: Thanks, done.

File: modules/grpc/common/grpc-source.hpp [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  void add_extra_channel_arg(std::string name, std::string value)
+  {
+    this->string_extra_channel_args.push_back(std::pair<std::string, std::string> {name, value});
```
[the remark this answers is not in the document]
Comment (alltilla) [reply modules/grpc/common/grpc-source.hpp]: Thanks, done.

**Review by OverOrion — APPROVED** [review]

File: modules/grpc/bigquery/bigquery-worker.cpp [hunk 15]
```diff
@@ -423,8 +423,8 @@ DestinationWorker::create_channel()
 {
   DestinationDriver *owner_ = this->get_owner();
 
-  auto credentials = ::grpc::GoogleDefaultCredentials();
   ::grpc::ChannelArguments args = this->create_channel_args();
+  auto credentials = this->create_credentials();
```
Comment (MrAnno) [inline modules/grpc/bigquery/bigquery-worker.cpp:427]: We shouldn't change the default here, BigQuery is most likely being used with `GCAM_ADC`.

**Comment by MrAnno** [comment]
I think we shouldn't introduce more dynamic libraries around the gRPC drivers, we already have one (static linking seems more than okay in this specific case).

File: modules/grpc/bigquery/bigquery-worker.cpp (hunk 15 above)
Comment (alltilla) [reply modules/grpc/bigquery/bigquery-worker.cpp:427]: Nice catch!

Added this to the ctor of `bigquery::DestinationDriver`:
```c++
this->credentials_builder.set_mode(GCAM_ADC);
```

**Comment by alltilla** [comment]
Fixed one review finding and rebased to main.

**Comment by MrAnno** [comment]
Let's ask @mitzkia to check the compatibility of this PR next week (manually or in an automated way).

**Comment by alltilla** [comment]
rebased to main

**Comment by alltilla** [comment]
Tested E2E with:
```
log {
  source { network(port(12345)); };

  # destination {
  #   opentelemetry(
  #     url("localhost:4317")
  #     batch-lines(10000)
  #     workers(4)
  #   );
  # };

  # destination {
  #   loki(
  #     url("localhost:9096")
  #     labels(
  #       "app" => "syslog-ng"
  #     )
  #     compression(yes)
  #     batch-bytes(10M)
  #   );
  # };
  # # curl -G \
  # #       --data-urlencode 'query=count_over_time({app="syslog-ng"}[1h])' \
  # #       http://localhost:3100/loki/api/v1/query | jq .data.stats.summary.totalLinesProcessed

  destination {
    bigquery(
      project("redacted-project")
      dataset("redacted_dataset")
      table("alltilla-test")
      batch-lines(10000)
      workers(4)
      schema (
        my_string STRING => $MESSAGE,
        my_int INTEGER => 42,
        my_bool BOOLEAN => "true",
      )
    );
  };

};
```

**Comment by alltilla** [comment]
@MrAnno I had to add some fixup commits that came from the manual E2E tests, could you kindly check them? Thanks!

File: modules/grpc/common/grpc-source.cpp [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+  if (!this->port)
+    {
+      msg_error("Failed to initialize gRPC based source, port() must be set",
+                log_pipe_location_tag(&this->super->super.super.super.super));
+      return FALSE;
+    }
+
+  ::grpc::EnableDefaultHealthCheckService(true);
+
+  this->super->super.worker_options.super.init_window_size /= super->super.num_workers;
```
Comment (MrAnno) [inline modules/grpc/common/grpc-source.cpp:50]: This is handled by threaded source now.

We should move this logic here in order not to revert #310 accidenteally:
```
 if (fetch_limit == -1)
    {
      if (super->super.worker_options.super.init_window_size != -1)
        fetch_limit = super->super.worker_options.super.init_window_size / super->super.num_workers;
      else
        fetch_limit = 100;
    }
```

**Review by MrAnno — CHANGES_REQUESTED** [review]
LGTM otherwise.

File: modules/grpc/common/grpc-source.cpp (hunk 16 above)
Comment (alltilla) [reply modules/grpc/common/grpc-source.cpp]: nice catch!

squashed the fixup commits and fixed `grpc::SourceDriver::init()` in its original commit,

**Review by MrAnno — APPROVED** [review]

**Pull request merged**